devtools/x/src/utils.rs

```diff
@@ -16,19 +21,102 @@ pub fn project_root() -> &'static Path {
         .unwrap()
 }
 
-pub fn locate_project(xctx: &XContext) -> Result<PathBuf> {
+pub fn locate_project(cargo_config: &CargoConfig) -> Result<PathBuf> {
     #[derive(Deserialize)]
     struct LocateProject {
         root: PathBuf,
     };
 
-    let output = Cargo::new(xctx.config().cargo_config(), "locate-project").run_with_output()?;
+    let output = Cargo::new(cargo_config, "locate-project", false).run_with_output()?;
     Ok(serde_json::from_slice::<LocateProject>(&output)?.root)
 }
 
-pub fn project_is_root(xctx: &XContext) -> Result<bool> {
-    let mut project = locate_project(xctx)?;
+pub fn project_is_root(cargo_config: &CargoConfig) -> Result<bool> {
+    let mut project = locate_project(cargo_config)?;
     project.pop();
 
     Ok(project == project_root())
 }
+
+fn stop_sccache_server() {
+    let mut sccache = Command::new("sccache");
+    sccache.arg("--stop-server");
+    let result = sccache.output();
+    if let Ok(output) = result {
+        if output.status.success() {
+            info!("Stopped already running sccache.");
+        }
+    }
+}
+
+pub fn apply_sccache_if_possible<'a>(
+    cargo_config: &'a CargoConfig,
+) -> Result<Vec<(&'a str, Option<&'a str>)>> {
+    let mut envs = vec![];
+
+    if option_env!("SKIP_SCCACHE").is_none() && cargo_config.sccache.is_some() {
+        if let Some(sccach_config) = &cargo_config.sccache {
+            // Are we work on items in the right location:
+            // See: https://github.com/mozilla/sccache#known-caveats
+            let correct_location = option_env!("CARGO_HOME").unwrap_or_default()
+                == sccach_config.required_cargo_home
+                && sccach_config.required_git_home == project_root().to_str().unwrap_or_default();
+            if !correct_location {
+                warn!("You will not benefit from sccache in this build!!!");
+                warn!(
+                    "To get the best experience, please move your libra source code to {} and your set your CARGO_HOME to be {}, simply export it in your .profile or .bash_rc",
+                    &sccach_config.required_git_home, &sccach_config.required_cargo_home
```

Comment:
Oh wow, had no idea sccache had such exacting requirements. Does this mean it's going to be hard for developers to set it up locally?

Reply:
Took me about 3-4 minutes.   Big thing is to ensure your IDE has the same CARGO_HOME.   I'll probably get around to adding that to the dev_setup.sh as well.